audio_downloader.py

```diff
@@ -0,0 +1,92 @@
+import logging
+import os
+from aiogram import types
+from aiogram.types import ParseMode
+from youtube_dl import YoutubeDL
+from urllib.parse import urlparse
+from urllib.request import Request
+from urllib.request import urlopen
+from PIL import Image
+import config
+
+logger = logging.getLogger(__name__)
+
+
+class AudioDownloader:
+
+    async def download(self, message: types.Message):
+        if not str(message.chat.id) in config.WHITELIST_CHAT_ID:
+            return
+        try:
+            ydl_opts = {
+                    'writethumbnail': True,
+                    'outtmpl': 'downloads/%(extractor)s_%(id)s.%(ext)s',
+                    'format': 'bestaudio',
+            }
+            ydl = YoutubeDL(ydl_opts)
+            info = ydl.extract_info(message.text, download=False)
+            audio_url = format(info['url'])
+            telegram_audio_limit = 52428800
+            request_head = Request(audio_url, method='HEAD')
+            audio_filesize = urlopen(request_head).headers['Content-Length']
+
+            if int(audio_filesize) < telegram_audio_limit:
+                ydl.download([message.text])
+                duration = format(info['duration'])
+                extractor = format(info['extractor'])
+                title = format(info['title'])
+                audio_ext = format(info['ext'])
+                webpage_id = format(info['id'])
+                webpage_url = format(info['webpage_url'])
+                uploader = format(info['uploader'])
+                thumbnail_url = format(info['thumbnail'])
+                thumbnail_urlpath = urlparse(thumbnail_url).path
+                thumbnail_filename = os.path.basename(thumbnail_urlpath)
+                thumbnail_ext = thumbnail_filename.split(".")[-1]
+                base_filename = "downloads/" + extractor + '_' + \
+                    webpage_id + '.'
+                output_audiofile = base_filename + audio_ext
+                original_thumbnail = base_filename + thumbnail_ext
+                output_thumbnail = "downloads/square_thumbnail_" + \
+                    webpage_id + ".jpg"
+                self.__make_squarethumb(original_thumbnail, output_thumbnail)
+                with open(output_audiofile, "rb") as audio, \
+                     open(output_thumbnail, "rb") as thumb:
+                    await message.reply_audio(audio,
```

Comment:
now the class that is supposed to be responsible for downloading an audio not only downloads audio but also replies to a message. this break a [single resposponsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)

you should try to make this **download** method return an entity (for example that entity could be called DownloadedAudio), by entity i mean a class that is going to be in this file which will have properties **audio**, **caption**, **webpage_url**, **title**, **duration**, **uploader**, **title** and **thumb**, or whatever is the best practice for it in python, maybe in python you have to use dictionary like you did with **info** for such a scenario.

The idea is you return this result from the **download** method and in **text_handler** if **AudioDownloader.download()** threw an exception you log the exception, if it's threw a valid result then you reply a message. So the code in text_handler will be something like
```
audio = AudioDownloader.download(...)
 message.reply_audio(audio.soundtrack, audio.caption, audio.webpage_url, audio.title..... etc 
```

Reply:
idk how to do this. I get a warning when I was trying that, seems that 'audio =' does not have 'await', but this can't be awaited.
>  RuntimeWarning: coroutine 'AudioDownloader.download' was never awaited

Reply:
I may could figure this out, will keep trying that tomorrow. but not sure how to deal with errors. **AudioDownloader** check audio file_size and decide to download it or not, but **TextHandler** can't interact in between. how to pass the error to **TextHandler**? Add an `error` entry to the dictionary which AudioDownloader returns and reply_audio when no "error" entry and reply text with content of "error" entry when there is?

Reply:
`how to pass the error to TextHandler`

if download(url) throws an error, then you'll catch this error in TextHandler

Reply:
yea ofc, await it and wrap up in a try/catch

Reply:
Review the latest commit